hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/SparkRangeBucketIndexPartitioner.java

```diff
@@ -0,0 +1,127 @@
+
+  public SparkRangeBucketIndexPartitioner(WorkloadProfile profile,
+                                          HoodieEngineContext context,
+                                          HoodieTable table,
+                                          HoodieWriteConfig config) {
+    super(profile, table);
+    if (!(table.getIndex() instanceof HoodieRangeBucketIndex)) {
+      throw new HoodieException(
+        " Bucket index partitioner should only be used by BucketIndex other than "
+          + table.getIndex().getClass().getSimpleName());
+    }
+    this.bucketRangeStepSize = config.getBucketRangeStepSize();
+    this.indexPartitionBucketNumMap = ((HoodieRangeBucketIndex) table.getIndex()).getIndexPartitionBucketNumMap();
+    this.partitionBucketNumIndexMap = ((HoodieRangeBucketIndex) table.getIndex()).getPartitionBucketNumIndexMap();
+
+    assignUpdates(profile);
+  }
+
+  private void assignUpdates(WorkloadProfile profile) {
+    updatePartitionPathFileIds = new HashMap<>();
+    // each update location gets a partition
+    Set<Entry<String, WorkloadStat>> partitionStatEntries = profile.getInputPartitionPathStatMap()
+        .entrySet();
+    for (Entry<String, WorkloadStat> partitionStat : partitionStatEntries) {
+      if (!updatePartitionPathFileIds.containsKey(partitionStat.getKey())) {
+        updatePartitionPathFileIds.put(partitionStat.getKey(), new HashSet<>());
+      }
+      for (Entry<String, Pair<String, Long>> updateLocEntry :
+          partitionStat.getValue().getUpdateLocationToCount().entrySet()) {
+        updatePartitionPathFileIds.get(partitionStat.getKey()).add(updateLocEntry.getKey());
+      }
+    }
+  }
+
+  @Override
+  public BucketInfo getBucketInfo(int partitionNum) {
+    Pair<String, Integer> pmod = indexPartitionBucketNumMap.get(partitionNum);
+    String partitionPath = pmod.getLeft();
+    String bucketId = BucketIdentifier.bucketIdStr(pmod.getRight());
+    Option<String> fileIdOption = Option.fromJavaOptional(updatePartitionPathFileIds
+        .getOrDefault(partitionPath, Collections.emptySet()).stream()
+        .filter(e -> e.startsWith(bucketId))
+        .findFirst());
+    if (fileIdOption.isPresent()) {
+      return new BucketInfo(BucketType.UPDATE, fileIdOption.get(), partitionPath);
+    } else {
+      return new BucketInfo(BucketType.INSERT, BucketIdentifier.newBucketFileIdPrefix(bucketId), partitionPath);
+    }
+  }
+
+  @Override
+  public int numPartitions() {
+    return indexPartitionBucketNumMap.size();
+  }
+
+  @Override
+  public int getPartition(Object key) {
+    Tuple2<HoodieKey, Option<HoodieRecordLocation>> keyLocation = (Tuple2<HoodieKey, Option<HoodieRecordLocation>>) key;
+    String partitionPath = keyLocation._1.getPartitionPath();
+
+    String recordKey = keyLocation._1.getRecordKey();
+    if (recordKey.contains(":")) {
+      recordKey = recordKey.substring(recordKey.indexOf(":") + 1);
+    }
+    long l = Long.parseLong(recordKey) / bucketRangeStepSize;
```

Comment:
Maybe extract the bucket number calculate logic to BucketIdentifier.

Reply:
done